File: agent/sdk/typescript/src/spot.ts
/**
 * Typed builders for every LayerX spot module activity (module 10). Each
 * encoder emits the exact bytes the kernel codec (`src/modules/spot`) accepts
 * and each decoder refuses what the kernel refuses.
 */

import {
  TradingPayloadError,
  TradingReader,
  TradingWriter,
  type EncodedActivity,
  type TradeSide,
  type TradingId,
} from "./perps.js";

export const SPOT_MODULE_ID = 10;

export const SPOT_ACTIVITY_TYPES = {
  market_create: 0x000a0001,
  order_place: 0x000a0002,
  order_cancel: 0x000a0003,
  market_halt: 0x000a0004,
  market_resume: 0x000a0005,
} as const;

/** 1 is a limit order, 2 a market order. */
export type SpotOrderKind = 1 | 2;

/** 1 is good-til-cancelled, 2 immediate-or-cancel. */
export type SpotTimeInForce = 1 | 2;

export type SpotActivity =
  | {
      readonly activity: "market_create";
      readonly marketId: TradingId;
      readonly baseAsset: TradingId;
      readonly quoteAsset: TradingId;
      readonly tickSize: bigint;
      readonly lotSize: bigint;
      readonly administrator: TradingId;
    }
  | {
      readonly activity: "order_place";
      readonly marketId: TradingId;
      readonly orderId: TradingId;
      readonly baseAccountId: TradingId;
      readonly quoteAccountId: TradingId;
      readonly side: TradeSide;
      readonly kind: SpotOrderKind;
      readonly timeInForce: SpotTimeInForce;
      readonly price: bigint;
      readonly quantity: bigint;
    }
  | { readonly activity: "order_cancel"; readonly marketId: TradingId; readonly orderId: TradingId }
  | { readonly activity: "market_halt"; readonly marketId: TradingId }
  | { readonly activity: "market_resume"; readonly marketId: TradingId };

function oneOrTwo(value: number, label: string): 1 | 2 {
  if (value !== 1 && value !== 2) {
    throw new TradingPayloadError("non_canonical", label);
  }
  return value;
}

/** Returns the packed activity type of a spot activity. */
export function spotActivityType(activity: SpotActivity): number {
  return SPOT_ACTIVITY_TYPES[activity.activity];
}

/** Encodes the exact kernel payload bytes of a spot activity. */
export function encodeSpotActivity(activity: SpotActivity): Uint8Array {
  const writer = new TradingWriter();
  switch (activity.activity) {
    case "market_create":
      if (
        activity.tickSize === 0n ||
        activity.lotSize === 0n ||
        activity.baseAsset === activity.quoteAsset ||
        /^0+$/u.test(activity.administrator)
      ) {
        throw new TradingPayloadError("non_canonical", "market");
      }
      writer
        .id(activity.marketId)
        .id(activity.baseAsset)
        .id(activity.quoteAsset)
        .uint(activity.tickSize, 16)
        .uint(activity.lotSize, 16)
        .id(activity.administrator);
      break;
    case "order_place": {
      const priced =
        activity.kind === 1 ? activity.price !== 0n : activity.price === 0n && activity.timeInForce === 2;
      if (!priced || activity.quantity === 0n || activity.baseAccountId === activity.quoteAccountId) {
        throw new TradingPayloadError("non_canonical", "order");
      }
      writer
        .id(activity.marketId)
        .id(activity.orderId)
        .id(activity.baseAccountId)
        .id(activity.quoteAccountId)
        .byte(oneOrTwo(activity.side, "side"))
        .byte(oneOrTwo(activity.kind, "kind"))
        .byte(oneOrTwo(activity.timeInForce, "time in force"))
        .uint(activity.price, 16)
        .uint(activity.quantity, 16);
      break;
    }
    case "order_cancel":
      writer.id(activity.marketId).id(activity.orderId);
      break;
    case "market_halt":
    case "market_resume":
      writer.id(activity.marketId);
      break;
  }
  return writer.bytes();
}

/** Builds the activity type and payload a spot activity submits. */
export function buildSpotActivity(activity: SpotActivity): EncodedActivity {
  return { activityType: spotActivityType(activity), payload: encodeSpotActivity(activity) };
}

const SPOT_LENGTHS: Readonly<Record<number, number>> = { 1: 160, 2: 163, 3: 64, 4: 32, 5: 32 };

/** Decodes and validates kernel payload bytes of a spot activity. */
export function decodeSpotActivity(activityType: number, payload: Uint8Array): SpotActivity {
  const ordinal = activityType & 0xffff;
  const expected = activityType >>> 16 === SPOT_MODULE_ID ? SPOT_LENGTHS[ordinal] : undefined;
  if (expected === undefined) {
    throw new TradingPayloadError("unknown_activity", activityType.toString(16));
  }
  if (payload.length !== expected) {
    throw new TradingPayloadError("length");
  }
  const reader = new TradingReader(payload);
  let decoded: SpotActivity;
  switch (ordinal) {
    case 1:
      decoded = {
        activity: "market_create",
        marketId: reader.id(),
        baseAsset: reader.id(),
        quoteAsset: reader.id(),
        tickSize: reader.uint(16),
        lotSize: reader.uint(16),
        administrator: reader.id(),
      };
      break;
    case 2:
      decoded = {
        activity: "order_place",
        marketId: reader.id(),
        orderId: reader.id(),
        baseAccountId: reader.id(),
        quoteAccountId: reader.id(),
        side: oneOrTwo(reader.byte(), "side"),
        kind: oneOrTwo(reader.byte(), "kind"),
        timeInForce: oneOrTwo(reader.byte(), "time in force"),
        price: reader.uint(16),
        quantity: reader.uint(16),
      };
      break;
    case 3:
      decoded = { activity: "order_cancel", marketId: reader.id(), orderId: reader.id() };
      break;
    case 4:
      decoded = { activity: "market_halt", marketId: reader.id() };
      break;
    default:
      decoded = { activity: "market_resume", marketId: reader.id() };
  }
  const encoded = encodeSpotActivity(decoded);
  if (encoded.length !== payload.length || encoded.some((byte, index) => byte !== payload[index])) {
    throw new TradingPayloadError("non_canonical");
  }
  return decoded;
}

File: agent/sdk/python/layerx_sdk/spot.py
"""Typed builders for every LayerX spot module activity (module 10).

Each encoder emits the exact bytes the kernel codec (``src/modules/spot``)
accepts and each decoder refuses what the kernel refuses.
"""

from __future__ import annotations

from dataclasses import dataclass

from .perps import TradingPayloadError, _Layout, decode_trading_activity

SPOT_MODULE_ID = 10
LIMIT = 1
MARKET = 2
GOOD_TIL_CANCELLED = 1
IMMEDIATE_OR_CANCEL = 2


class _Spot(_Layout):
    MODULE = SPOT_MODULE_ID


@dataclass(frozen=True)
class SpotMarketCreate(_Spot):
    market_id: str
    base_asset: str
    quote_asset: str
    tick_size: int
    lot_size: int
    administrator: str
    ORDINAL = 1
    LAYOUT = ("id", "id", "id", "u128", "u128", "id")

    def check(self) -> None:
        if self.tick_size == 0 or self.lot_size == 0 or self.base_asset == self.quote_asset:
            raise TradingPayloadError("non_canonical", "market")


@dataclass(frozen=True)
class SpotOrderPlace(_Spot):
    market_id: str
    order_id: str
    base_account_id: str
    quote_account_id: str
    side: int
    kind: int
    time_in_force: int
    price: int
    quantity: int
    ORDINAL = 2
    LAYOUT = ("id", "id", "id", "id", "choice", "choice", "choice", "u128", "u128")

    def check(self) -> None:
        priced = self.price != 0 if self.kind == LIMIT else self.price == 0 and self.time_in_force == IMMEDIATE_OR_CANCEL
        if not priced or self.quantity == 0 or self.base_account_id == self.quote_account_id:
            raise TradingPayloadError("non_canonical", "order")


@dataclass(frozen=True)
class SpotOrderCancel(_Spot):
    market_id: str
    order_id: str
    ORDINAL = 3
    LAYOUT = ("id", "id")


@dataclass(frozen=True)
class SpotMarketHalt(_Spot):
    market_id: str
    ORDINAL = 4
    LAYOUT = ("id",)


@dataclass(frozen=True)
class SpotMarketResume(_Spot):
    market_id: str
    ORDINAL = 5
    LAYOUT = ("id",)


SpotActivity = SpotMarketCreate | SpotOrderPlace | SpotOrderCancel | SpotMarketHalt | SpotMarketResume

SPOT_ACTIVITIES: dict[int, type[_Layout]] = {
    kind.ORDINAL: kind for kind in (SpotMarketCreate, SpotOrderPlace, SpotOrderCancel, SpotMarketHalt, SpotMarketResume)
}


def build_spot_activity(activity: SpotActivity) -> tuple[int, bytes]:
    """The packed activity type and exact payload bytes a spot activity submits."""
    return activity.activity_type, activity.encode()


def decode_spot_activity(activity_type: int, payload: bytes) -> SpotActivity:
    """Decodes and validates kernel payload bytes of a spot activity."""
    return decode_trading_activity(SPOT_ACTIVITIES, SPOT_MODULE_ID, activity_type, payload)  # type: ignore[return-value]
